Add CSS unique selectors to shipping methods names on the backend and hook after shipping zones table

webdados opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Add CSS unique selectors to shipping methods names on the backend and hook after shipping zones table

Closes #51816

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

1. Check that the Shipping methods titles (LI) on the shipping zones main screen now have a data-id attribute
2. Hook into the new action woocommerce_shipping_zones_after_zones_table and add content after the table (for example, css targeting each shipping method by data-id

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message

Add CSS unique selectors to shipping methods names on the backend and hook after shipping zones table

</details>

Commit:
Add CSS unique selectors to shipping methods names on the backend and hook after shipping zones table

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

naman03malhotra commented on plugins/woocommerce/client/legacy/js/admin/wc-shipping-zones.js:
```diff
@@ -169,7 +169,7 @@
 
 							$method_list.append(
-								'<li class="wc-shipping-zone-method ' + class_name + '">' + shipping_method.title + '</li>'
+								'<li class="wc-shipping-zone-method ' + class_name + '" data-id="' + shipping_method.instance_id + '">' + shipping_method.title + '</li>'
```
I agree this can be more helpful for styling. But Alternatively, have you considered [nth_child](https://www.w3schools.com/cssref/sel_nth-child.php) CSS for styling it?

webdados replied:
nth_child will not allow me to target a specific shipping method.

naman03malhotra commented on plugins/woocommerce/includes/admin/settings/views/html-admin-page-shipping-zones.php:
```diff
@@ -65,6 +65,11 @@
 
+<?php
+// phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingHookComment
+do_action( 'woocommerce_shipping_zones_after_zones_table' );
```
As I understood, the intent of this hook is that you would like to display the zones table?

webdados replied:
The intent is to have a hook in which to be able to inject extra styling per shipping method. In my specific case, I'll be adding an icon.

kloon replied:
@webdados Can you please add a hook comment and remove the phpcs:ignore.

webdados replied:
Done: https://github.com/woocommerce/woocommerce/pull/51817/commits/40d365dfd23e6d40ebf3622f8e642a52f9cfa1ab

kloon commented on plugins/woocommerce/includes/admin/settings/views/html-admin-page-shipping-zones.php:
```diff
@@ -65,6 +65,11 @@
 
+<?php
+// phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingHookComment
+do_action( 'woocommerce_shipping_zones_after_zones_table' );
```
```suggestion
do_action( 'woocommerce_shipping_zones_after_methods_table' );
```
I'd suggest to change the name as above as this is after the shipping methods table on the shipping zones page.

kloon replied:
I am also curious about this hook. What would you like to add to this page via the hook? We aim to keep these pages focused on their intentions, so adding a general hook like this could mean the page could become cluttered with other things that are not necessarily relevant to shipping zone methods.

webdados replied:
Hi @kloon and thanks for your interest in this PR.

The shipping methods list inside each zone (Settings > Shipping > Shipping zones > Inside a zone) already has a data-id argument, which allows for direct css or js targeting, but the same does not happen on the general view in Settings > Shipping > Shipping zones. This PR intends to make this coherent on both views.

As for my use case, I have a plugin which allows for pickup location selection on the frontend. On this plugin settings, the shop owner will select which shipping methods will be eligible for that.

<img width="573" alt="image" src="https://github.com/user-attachments/assets/eaaf0a39-4b0f-45a4-a23b-2b0f9b86d68a" />

Our customers (shop owner) have been giving us feedback that it would be useful to have some king of feedback on the shipping zones settings so that they know which methods are selected on our plugin. This PR allows us to target those LI elements and style it:

<img width="1604" alt="image" src="https://github.com/user-attachments/assets/35a7d952-106f-4a1d-aee8-fcc83c63f044" />

I will take care of removing the phpcs comment and add a proper hook comment. Sorry for that :-D

webdados replied:
> I'd suggest to change the name as above as this is after the shipping methods table on the shipping zones page.


The methods are listed on each table row, but this is actually the shipping zones table, so I think the hook name should not be changed as proposed.

<img width="413" alt="image" src="https://github.com/user-attachments/assets/8eeab80b-eb31-4c8f-be8b-fc9353e3a8e9" />

webdados replied:
> I'd suggest to change the name as above as this is after the shipping methods table on the shipping zones page.

The methods are listed on each zone row, but the table itself is a zones table, not a methods table (which resides inside each zone details), so I don't think the proposed change on the hook name should be done.

<img width="524" alt="image" src="https://github.com/user-attachments/assets/b5350c01-4a9d-4a72-8b92-55061a8e4eb3" />

webdados replied:
phpcs comment removed and hook comment added

https://github.com/woocommerce/woocommerce/pull/51817/commits/40d365dfd23e6d40ebf3622f8e642a52f9cfa1ab

webdados committed:
Add woocommerce_shipping_zones_after_zones_table hook comment

Add woocommerce_shipping_zones_after_zones_table hook comment instead of phpcs:ignore

The @since parameter was not added as it is not known when this will be merged and shipped into stable.

webdados commented:
In the meantime @psealock did #54823 which creates the data-id attributes.
There goes my name as a contributor for next release 😂

Should I cancel this PR and create another one just with the new proposed hook after the shipping zones table?

kloon commented:
Great catch, if that is not too much effort then I would suggest it yes. Feel free to request a review from me directly if you can, otherwise just drop a link to the new PR here and I will pick it up.

psealock commented:
>The intent is to have a hook in which to be able to inject extra styling per shipping method. In my specific case, I'll be adding an icon.

If you are just adding CSS to the page with this hook, does it matter where its printed? I'm just wondering if making use of [`woocommerce_before_settings_shipping`](https://github.com/woocommerce/woocommerce/blob/2d65567a6ad261a043a95b4f758f72ed190fd0de/plugins/woocommerce/includes/admin/views/html-admin-settings.php#L27) will suffice?

webdados commented:
@psealock Yes, it suffices. I was adding the other one to be consistent with the table inside each zone, which already has those kind of hooks. Also hooking into those specific ones would guarantee that I would be running my code (because I do have to go through my plugin settings to generate CSS dynamically only for those IDs that match our settings) on the correct pages, and not on every shipping settings pages (with woocommerce_before_settings_shipping). Anyway, I can just look into the GET params and run my code only when needed.

Thanks for bringing this up.

I will close this PR as I already have everything I need in place.
